File: sketch/src/main/java/com/github/panpf/sketch/util/pool/SizeConfigStrategy.java
package com.github.panpf.sketch.util.pool;

import android.graphics.Bitmap;
import android.os.Build;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;

import com.github.panpf.sketch.util.BitmapUtilsKt;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import java.util.NavigableMap;
import java.util.Objects;
import java.util.TreeMap;

/**
 * Keys {@link Bitmap Bitmaps} using both {@link Bitmap#getAllocationByteCount()} and
 * the {@link Bitmap.Config} returned from {@link Bitmap#getConfig()}.
 * <p>
 * <p>
 * Using both the config and the byte size allows us to safely re-use a greater variety of
 * {@link Bitmap Bitmaps}, which increases the hit rate of the pool and therefore the performance
 * of applications. This class works around #301 by only allowing re-use of {@link Bitmap Bitmaps}
 * with a matching number of bytes per pixel.
 * </p>
 */
@RequiresApi(Build.VERSION_CODES.KITKAT)
public class SizeConfigStrategy implements LruPoolStrategy {

    private static final int MAX_SIZE_MULTIPLE = 8;
    @NonNull
    private static final Bitmap.Config[] ARGB_8888_IN_CONFIGS;

    static {
        // null: The value returned by Bitmaps with the hidden Bitmap config.
        Bitmap.Config[] result = new Bitmap.Config[]{Bitmap.Config.ARGB_8888, null};
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            result = Arrays.copyOf(result, result.length + 1);
            result[result.length - 1] = Bitmap.Config.RGBA_F16;
        }
        ARGB_8888_IN_CONFIGS = result;
    }

    @NonNull
    private static final Bitmap.Config[] RGBA_F16_IN_CONFIGS = ARGB_8888_IN_CONFIGS;

    // We probably could allow ARGB_4444 and RGB_565 to decode into each other, but ARGB_4444 is
    // deprecated and we'd rather be safe.
    @NonNull
    private static final Bitmap.Config[] RGB_565_IN_CONFIGS =
            new Bitmap.Config[]{Bitmap.Config.RGB_565};
    @NonNull
    private static final Bitmap.Config[] ARGB_4444_IN_CONFIGS =
            new Bitmap.Config[]{Bitmap.Config.ARGB_4444};
    @NonNull
    private static final Bitmap.Config[] ALPHA_8_IN_CONFIGS =
            new Bitmap.Config[]{Bitmap.Config.ALPHA_8};

    @NonNull
    private final KeyPool keyPool = new KeyPool();
    @NonNull
    private final GroupedLinkedMap<Key, Bitmap> groupedMap = new GroupedLinkedMap<>();
    @NonNull
    private final Map<Bitmap.Config, NavigableMap<Integer, Integer>> sortedSizes = new HashMap<>();

    @Override
    public void put(@NonNull Bitmap bitmap) {
        int size = BitmapUtilsKt.getAllocationByteCountCompat(bitmap);
        Key key = keyPool.get(size, bitmap.getConfig());

        groupedMap.put(key, bitmap);

        NavigableMap<Integer, Integer> sizes = getSizesForConfig(bitmap.getConfig());
        Integer current = sizes.get(key.size);
        sizes.put(key.size, current == null ? 1 : current + 1);
    }

    @Override
    public Bitmap get(int width, int height, @Nullable Bitmap.Config config) {
        int size = BitmapUtilsKt.getBitmapByteSize(width, height, config);
        Key bestKey = findBestKey(size, config);

        Bitmap result = groupedMap.get(bestKey);
        if (result != null) {
            // Decrement must be called before reconfigure.
            decrementBitmapOfSize(bestKey.size, result);
            try {
                result.reconfigure(width, height, config);
            } catch (IllegalArgumentException e) {
                // Bitmap.cpp Bitmap_reconfigure method may throw "IllegalArgumentException: Bitmap not large enough to support new configuration" exception
                e.printStackTrace();
                put(result);
            }
        }
        return result;
    }

    @Override
    public boolean exist(int width, int height, @Nullable Bitmap.Config config) {
        int size = BitmapUtilsKt.getBitmapByteSize(width, height, config);
        Key bestKey = findBestKey(size, config);
        return groupedMap.exist(bestKey);
    }

    @NonNull
    private Key findBestKey(int size, @Nullable Bitmap.Config config) {
        Key result = keyPool.get(size, config);
        for (Bitmap.Config possibleConfig : getInConfigs(config)) {
            NavigableMap<Integer, Integer> sizesForPossibleConfig = getSizesForConfig(possibleConfig);
            Integer possibleSize = sizesForPossibleConfig.ceilingKey(size);
            if (possibleSize != null && possibleSize <= size * MAX_SIZE_MULTIPLE) {
                if (possibleSize != size || (!Objects.equals(possibleConfig, config))) {
                    keyPool.offer(result);
                    result = keyPool.get(possibleSize, possibleConfig);
                }
                break;
            }
        }
        return result;
    }

    @Nullable
    @Override
    public Bitmap removeLast() {
        Bitmap removed = groupedMap.removeLast();
        if (removed != null) {
            int removedSize = BitmapUtilsKt.getAllocationByteCountCompat(removed);
            decrementBitmapOfSize(removedSize, removed);
        }
        return removed;
    }

    private void decrementBitmapOfSize(@NonNull Integer size, @NonNull Bitmap removed) {
        Bitmap.Config config = removed.getConfig();
        NavigableMap<Integer, Integer> sizes = getSizesForConfig(config);
        Integer current = sizes.get(size);
        if (current == null) {
            String message = String.format(
                    Locale.getDefault(),
                    "Tried to decrement empty size, size: %d, removed: %s, this: %s",
                    size, logBitmap(removed), this
            );
            throw new NullPointerException(message);
        }
        if (current == 1) {
            sizes.remove(size);
        } else {
            sizes.put(size, current - 1);
        }
    }

    @NonNull
    private NavigableMap<Integer, Integer> getSizesForConfig(@Nullable Bitmap.Config config) {
        NavigableMap<Integer, Integer> sizes = sortedSizes.get(config);
        if (sizes == null) {
            sizes = new TreeMap<>();
            sortedSizes.put(config, sizes);
        }
        return sizes;
    }

    @NonNull
    @Override
    public String logBitmap(@NonNull Bitmap bitmap) {
        int size = BitmapUtilsKt.getAllocationByteCountCompat(bitmap);
        return getBitmapString(size, bitmap.getConfig());
    }

    @NonNull
    @Override
    public String logBitmap(int width, int height, @Nullable Bitmap.Config config) {
        int size = BitmapUtilsKt.getBitmapByteSize(width, height, config);
        return getBitmapString(size, config);
    }

    @Override
    public int getSize(@NonNull Bitmap bitmap) {
        return BitmapUtilsKt.getAllocationByteCountCompat(bitmap);
    }

    @NonNull
    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder()
                .append("SizeConfigStrategy(groupedMap=")
                .append(groupedMap)
                .append(", sortedSizes=(");
        for (Map.Entry<Bitmap.Config, NavigableMap<Integer, Integer>> entry : sortedSizes.entrySet()) {
            sb.append(entry.getKey()).append('[').append(entry.getValue()).append("], ");
        }
        if (!sortedSizes.isEmpty()) {
            sb.replace(sb.length() - 2, sb.length(), "");
        }
        return sb.append("))").toString();
    }

    // Visible for testing.
    static class KeyPool extends BaseKeyPool<Key> {

        @NonNull
        public Key get(int size, @Nullable Bitmap.Config config) {
            Key result = get();
            result.init(size, config);
            return result;
        }

        @NonNull
        @Override
        protected Key create() {
            return new Key(this);
        }
    }

    // Visible for testing.
    static final class Key implements Poolable {

        @NonNull
        private final KeyPool pool;
        private int size;
        @Nullable
        private Bitmap.Config config;

        public Key(@NonNull KeyPool pool) {
            this.pool = pool;
        }

        public void init(int size, @Nullable Bitmap.Config config) {
            this.size = size;
            this.config = config;
        }

        @Override
        public void offer() {
            pool.offer(this);
        }

        @NonNull
        @Override
        public String toString() {
            return getBitmapString(size, config);
        }

        @Override
        public boolean equals(Object o) {
            if (o instanceof Key) {
                Key other = (Key) o;
                return size == other.size && (Objects.equals(config, other.config));
            }
            return false;
        }

        @Override
        public int hashCode() {
            int result = size;
            result = 31 * result + (config != null ? config.hashCode() : 0);
            return result;
        }
    }

    @NonNull
    private static String getBitmapString(int size, @Nullable Bitmap.Config config) {
        return "[" + size + "](" + config + ")";
    }

    @NonNull
    private static Bitmap.Config[] getInConfigs(@Nullable Bitmap.Config requested) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            if (Bitmap.Config.RGBA_F16.equals(requested)) { // NOPMD - Avoid short circuiting sdk checks.
                return RGBA_F16_IN_CONFIGS;
            }
        }

        if (requested == null) {
            return new Bitmap.Config[]{null};
        }
        switch (requested) {
            case ARGB_8888:
                return ARGB_8888_IN_CONFIGS;
            case RGB_565:
                return RGB_565_IN_CONFIGS;
            case ARGB_4444:
                return ARGB_4444_IN_CONFIGS;
            case ALPHA_8:
                return ALPHA_8_IN_CONFIGS;
            default:
                return new Bitmap.Config[]{requested};
        }
    }
}

File: sketch-core/src/main/kotlin/com/github/panpf/sketch/util/pool/SizeConfigStrategy.kt
@file:Suppress("DEPRECATION")

package com.github.panpf.sketch.util.pool

import android.graphics.Bitmap
import android.graphics.Bitmap.Config
import android.graphics.Bitmap.Config.ALPHA_8
import android.graphics.Bitmap.Config.ARGB_4444
import android.graphics.Bitmap.Config.ARGB_8888
import android.graphics.Bitmap.Config.RGBA_F16
import android.graphics.Bitmap.Config.RGB_565
import android.os.Build.VERSION
import android.os.Build.VERSION_CODES
import androidx.annotation.RequiresApi
import com.github.panpf.sketch.decode.internal.logString
import com.github.panpf.sketch.util.allocationByteCountCompat
import com.github.panpf.sketch.util.calculateBitmapByteCount
import java.util.Locale
import java.util.NavigableMap
import java.util.TreeMap

/**
 * Keys [Bitmaps][Bitmap] using both [Bitmap.getAllocationByteCount] and
 * the [Bitmap.Config] returned from [Bitmap.getConfig].
 *
 *
 *
 *
 * Using both the config and the byte size allows us to safely re-use a greater variety of
 * [Bitmaps][Bitmap], which increases the hit rate of the pool and therefore the performance
 * of applications. This class works around #301 by only allowing re-use of [Bitmaps][Bitmap]
 * with a matching number of bytes per pixel.
 *
 */
@RequiresApi(VERSION_CODES.KITKAT)
class SizeConfigStrategy : LruPoolStrategy {

    private val keyPool = KeyPool()
    private val groupedMap = GroupedLinkedMap<Key, Bitmap>()
    private val sortedSizes: MutableMap<Config?, NavigableMap<Int, Int>?> = HashMap()

    override fun put(bitmap: Bitmap) {
        val size = bitmap.allocationByteCountCompat
        val key = keyPool[size, bitmap.config]
        groupedMap.put(key, bitmap)
        val sizes = getSizesForConfig(bitmap.config)
        val current = sizes[key.size]
        sizes[key.size] = if (current == null) 1 else current + 1
    }

    override operator fun get(width: Int, height: Int, config: Config): Bitmap? {
        val size = calculateBitmapByteCount(width, height, config)
        val bestKey = findBestKey(size, config)
        var result = groupedMap[bestKey]
        if (result != null) {
            // Decrement must be called before reconfigure.
            decrementBitmapOfSize(bestKey.size, result)
            try {
                result.reconfigure(width, height, config)
            } catch (e: IllegalArgumentException) {
                // Bitmap.cpp Bitmap_reconfigure method may throw "IllegalArgumentException: Bitmap not large enough to support new configuration" exception
                val message = String.format(
                    Locale.getDefault(),
                    "Bitmap reconfigure error. size=%dx%d, config=%s, bitmap=%s",
                    width, height, config, result.logString
                )
                IllegalArgumentException(message, e).printStackTrace()
                put(result)
                result = null
            }
        }
        return result
    }

    override fun exist(width: Int, height: Int, config: Config): Boolean {
        val size = calculateBitmapByteCount(width, height, config)
        val bestKey = findBestKey(size, config)
        return groupedMap.exist(bestKey)
    }

    private fun findBestKey(size: Int, config: Config?): Key {
        var result = keyPool[size, config]
        for (possibleConfig in getInConfigs(config)) {
            val sizesForPossibleConfig = getSizesForConfig(possibleConfig)
            val possibleSize = sizesForPossibleConfig.ceilingKey(size)
            if (possibleSize != null && possibleSize <= size * MAX_SIZE_MULTIPLE) {
                if (possibleSize != size || possibleConfig != config) {
                    keyPool.offer(result)
                    result = keyPool[possibleSize, possibleConfig]
                }
                break
            }
        }
        return result
    }

    override fun removeLast(): Bitmap? {
        val removed = groupedMap.removeLast()
        if (removed != null) {
            val removedSize = removed.allocationByteCountCompat
            decrementBitmapOfSize(removedSize, removed)
        }
        return removed
    }

    private fun decrementBitmapOfSize(size: Int, removed: Bitmap) {
        val config = removed.config
        val sizes = getSizesForConfig(config)
        val current = sizes[size]
        if (current == null) {
            val message = String.format(
                Locale.getDefault(),
                "Tried to decrement empty size, size: %d, removed: %s, this: %s",
                size, logBitmap(removed), this
            )
            throw NullPointerException(message)
        }
        if (current == 1) {
            sizes.remove(size)
        } else {
            sizes[size] = current - 1
        }
    }

    private fun getSizesForConfig(config: Config?): NavigableMap<Int, Int> {
        var sizes = sortedSizes[config]
        if (sizes == null) {
            sizes = TreeMap()
            sortedSizes[config] = sizes
        }
        return sizes
    }

    override fun logBitmap(bitmap: Bitmap): String {
        val size = bitmap.allocationByteCountCompat
        return getBitmapString(size, bitmap.config)
    }

    override fun logBitmap(width: Int, height: Int, config: Config): String {
        val size = calculateBitmapByteCount(width, height, config)
        return getBitmapString(size, config)
    }

    override fun getSize(bitmap: Bitmap): Int {
        return bitmap.allocationByteCountCompat
    }

    override fun toString(): String {
        val sb = StringBuilder()
            .append("SizeConfigStrategy(groupedMap=")
            .append(groupedMap)
            .append(", sortedSizes=(")
        for ((key, value) in sortedSizes) {
            sb.append(key).append('[').append(value).append("], ")
        }
        if (sortedSizes.isNotEmpty()) {
            sb.replace(sb.length - 2, sb.length, "")
        }
        return sb.append("))").toString()
    }

    // Visible for testing.
    internal class KeyPool : BaseKeyPool<Key?>() {
        operator fun get(size: Int, config: Config?): Key {
            val result = get()
            result!!.init(size, config)
            return result
        }

        override fun create(): Key {
            return Key(this)
        }
    }

    // Visible for testing.
    internal class Key(private val pool: KeyPool) : Poolable {

        var size = 0
        private var config: Config? = null

        fun init(size: Int, config: Config?) {
            this.size = size
            this.config = config
        }

        override fun offer() {
            pool.offer(this)
        }

        override fun toString(): String {
            return getBitmapString(size, config)
        }

        override fun equals(other: Any?): Boolean {
            if (other is Key) {
                return size == other.size && config == other.config
            }
            return false
        }

        override fun hashCode(): Int {
            var result = size
            result = 31 * result + if (config != null) config.hashCode() else 0
            return result
        }
    }

    companion object {
        private const val MAX_SIZE_MULTIPLE = 8
        private val ARGB_8888_IN_CONFIGS: Array<Config?>
        private val RGBA_F16_IN_CONFIGS: Array<Config?>

        // We probably could allow ARGB_4444 and RGB_565 to decode into each other, but ARGB_4444 is
        // deprecated and we'd rather be safe.
        private val RGB_565_IN_CONFIGS = arrayOf<Config?>(RGB_565)
        private val ARGB_4444_IN_CONFIGS = arrayOf<Config?>(ARGB_4444)
        private val ALPHA_8_IN_CONFIGS = arrayOf<Config?>(ALPHA_8)

        init {
            // null: The value returned by Bitmaps with the hidden Bitmap config.
            var result = arrayOf(ARGB_8888, null)
            if (VERSION.SDK_INT >= VERSION_CODES.O) {
                result = result.copyOf(result.size + 1)
                result[result.size - 1] = RGBA_F16
            }
            ARGB_8888_IN_CONFIGS = result
            RGBA_F16_IN_CONFIGS = result
        }

        private fun getBitmapString(size: Int, config: Config?): String {
            return "[$size]($config)"
        }

        private fun getInConfigs(requested: Config?): Array<Config?> {
            if (VERSION.SDK_INT >= VERSION_CODES.O) {
                if (RGBA_F16 == requested) { // NOPMD - Avoid short circuiting sdk checks.
                    return RGBA_F16_IN_CONFIGS
                }
            }
            return if (requested == null) {
                arrayOf(null)
            } else when (requested) {
                ARGB_8888 -> ARGB_8888_IN_CONFIGS
                RGB_565 -> RGB_565_IN_CONFIGS
                ARGB_4444 -> ARGB_4444_IN_CONFIGS
                ALPHA_8 -> ALPHA_8_IN_CONFIGS
                else -> arrayOf(requested)
            }
        }
    }
}